network/src/network.rs

```diff
@@ -331,12 +339,17 @@ impl NetworkState {
             .collect::<Vec<_>>()
     }
 
-    pub(crate) fn can_dial(&self, peer_id: &PeerId, addr: &Multiaddr) -> bool {
-        if self.local_peer_id() == peer_id {
+    pub(crate) fn can_dial(
+        &self,
+        peer_id: &PeerId,
+        addr: &Multiaddr,
+        allow_dial_to_self: bool,
+    ) -> bool {
+        if !allow_dial_to_self && self.local_peer_id() == peer_id {
             trace!("Do not dial self: {:?}, {}", peer_id, addr);
             return false;
         }
-        if self.listened_addresses.read().contains_key(&addr) {
+        if self.listened_addrs.read().contains_key(&addr) {
```

Comment:
Maybe it will have problems if we remove `original_listened_addrs` and just add the binded addr to `listened_addrs`.